### DISCUSSION TITLE
Removing nodes from graph (and corresponding node features) and updating edge indices

### DISCUSSION THREAD

**Discussion by amjass12** [pyg-team/pytorch_geometric#6644] [Q&A]
Hi!

I have trained a GAT which works fine and I am now trying to do run predictions on on subgraphs of the main graph - the task being, remove a node and its node feature and re-run through model, each time getting the updated embeddings of the first gat_conv layer with attention values. 

I am getting a `Found indices in 'edge_index' that are larger than`... error which I understand but am unsure about how to fix.

My code to remove a node and its node features which produce the error above is:

```
to the original edges_index:
edges_numpy = edge_index.numpy()
#to remove node 0
edge_index_source = list(np.where(edges_numpy[0] == 0))
edge_index_target = list(np.where(edges_numpy[1] == 0))

new_adj = np.delete(edges_numpy, edge_index_source, axis=1)
new_adj = np.delete(new_adj, edge_index_target, axis=1)

new_tensor = torch.tensor(new_adj)
idx = 0
new_node_features = np.delete(final_node_features, idx, 0)

model.gat_conv1(new_node_features, new_tensor, return_attention_weights=True)

ValueError: Found indices in 'edge_index' that are larger than 3514 (got 3515). Please ensure that all indices in 'edge_index' point to valid indices in the interval [0, 3515) in your node feature matrix and try again.
```

Is there a straightforward way to fix this error (note also I am trying to do this in original tensor and not create a new dataset each time as it's an iterative exercise. 

Any help is appreciated. Thank you!

**Comment by wsad1** [comment]
The problem node ids in the `edge_index` need to be relabeled after removing a node. 
Trying using the [subgraph](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.subgraph).
```
subset = torch.ones_like(num_nodex, dtype = bool)
subset[node_id_to_remove] = False
new_edge_index = subgraph(subset, relabel_nodes = True)
new_node_feature = node_feature[subset]
```

**Reply by amjass12** [reply]
Hi @wsad1 ,

thank you so much for your reply. the `new_node_feature = node_feature[subset]` works without issue. but the subgraph functions throws an error requiring the edge_index_argument. This makes sense, however, i am not sure if the implementation is right. I have modified as follows:

```
subset = torch.ones_like(range(3516), dtype = bool)#3516 is number of nodes in data
subset[0] = False
new_edge_index = subgraph(edges_index, subset, relabel_nodes = True)
#this now works however not sure if this is the intended behaviour?
```

The new_edge_index now returns a tuple. the first element being the edge index and the second returning None. this does not seem right to me? Another issue is that, as they have been re-indexed the new 'node 0' no longer corresponds to the 'real' node 0 that was just removed. im slightly confused by this. would i have to re-index the nodes on the networkx side so that i am keep track of which nodes are being affected. also does this affect the GAT, i.e. the GAT does not now expect node 0 to be the original node 0 correct?

also... if i am doing an iterative node removal process. would the next step has a False mask **ONLY** at the next node removed? or would i have to mask node 0 as well again?

Many thanks!

**Reply by wsad1** [reply]
You're right we need to pass `edge_index` to `subgraph`. `edge_index, _ = subgraph(edges_index, subset, relabel_nodes = True)`, its returning a tuple of `edge_index` and `edge_attr` which is `None` as you saw.

You don't need to re-index the node. The `GAT` is agnostic to the index (permutation) of the node, so that's not an issue. 

Yes for an iterative node removal process, you only need to set to `False` the next node removed, because the previous node isn't part of the graph anymore.

**Reply by amjass12** [reply]
Hi @wsad1 , 

excellent, thank you (yes i saw shortly after I messaged you - the source code for subgraph and saw that it returns the `edge_attr` as default (should have read that first).

'You don't need to re-index the node. The GAT is agnostic to the index (permutation) of the node, so that's not an issue.'

The thing that is confusing me, in my mind is that the edge_index corresponds uniquely to the nodes on the graph (and this is where i am misunderstanding) - is that if i remove node 0 from the graph with all its edges in the original index, all node 'ones' now generate a new index of 0 - and so on and so forth - Is the GAT invariant to this indexing because:

**The node features only attend to their neighbouring nodes and are masked from nodes they do not connect to on the graph - so as long as the node feature index matches the new subgraph (edge_index) index - then this would satisfy this property?**

and so - for an iterative node removal: is the following correct for removing 2 nodes (node 0, and node 1 for simplicity)

```
#remove node zero
subset = torch.ones_like(torch.tensor(range(torch_graph.num_nodes)), dtype = bool)
subset[0] = False
new_edge_index, _ = subgraph(edge_index=torch.tensor(torch_graph.edge_index.numpy()), subset=subset, relabel_nodes=True)
new_node_features = final_node_features[subset]

#remove node one
subset = torch.ones_like(torch.tensor(range(torch_graph.num_nodes)), dtype = bool)
subset[1 - 1] = False #because node 1 - 1 (or any other node number) is node zero in the new edge_index from before

new_edge_index, _ = subgraph(edge_index=torch.tensor(new_edge_index.numpy()), subset=subset, relabel_nodes=True)
new_edge_index

subset_nf = torch.ones_like(torch.tensor(range(new_node_features.shape[0])), dtype = bool)
subset_nf[1-1] = False
new_node_features = new_node_features[subset_nf]

```
and this way - the index of the node features should always match the correct node indexes of the edges. Also, if i know that node features have changed as a result of node removal, I am free to change those feature values after node removal right? thank you!

**Reply by wsad1** [reply]
Since you are setting `relabel_nodes = True` the `edge_index` gets reindexed. 
The code you shared above looks right to me.

**Reply by amjass12** [reply]
@wsad1 , 

thank you very very much for all of your help, really appreciate it!
